src/searchmultipleEndpoint.js

```diff
@@ -0,0 +1,149 @@
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+    ***** END LICENSE BLOCK *****
+*/
+
+const config = require('config');
+const Translate = require('./translation/translate');
+const TextSearch = require('./textSearch');
+
+const SELECT_TIMEOUT = 60;
+const sessionsWaitingForSelection = {};
+
+var requestsSinceGC = 0;
+
+var SearchMultipleEndpoint = module.exports = {
+	handle: async function (ctx, next) {
+		ctx.assert(ctx.is('text/plain') || ctx.is('json'), 415);
+
+        setTimeout(() => {
+            gc();
+        });
+
+
+		var data = ctx.request.body;
+
+		if (!data) {
+			ctx.throw(400, "POST data not provided\n");
+		}
+
+        // If follow-up request, retrieve session and update context
+		var query;
+		var session;
+		if (typeof data == 'object') {
+			let sessionID = data.session;
+			if (!sessionID) {
+				ctx.throw(400, "'session' not provided");
+			}
+			session = sessionsWaitingForSelection[sessionID];
+			if (session) {
+				delete sessionsWaitingForSelection[sessionID];
+				session.ctx = ctx;
+				session.next = next;
+				session.data = data;
+			} else {
+				let single = !!ctx.request.query.single;
+				session = new SearchMultipleSession(ctx, next, data, { single });
+			}
+		}
+		else {
+            // Look for DOI, ISBN, etc.
+            var identifiers = Zotero.Utilities.Internal.extractIdentifiers(data);
+			let single = !!ctx.request.query.single;
+			session = new SearchMultipleSession(ctx, next, data, { single });
+		}
+
+        await session.handleSearchMultiple();
+
+        if (ctx.response.status == 300) {
+			if(typeof data == 'object') {
+				// Select item if this was an item selection query
+				session.data = data;
+				await session.handleSearchMultiple();
+			} else {
+				// Store session if returning multiple choices
+				sessionsWaitingForSelection[session.id] = session;
+			}
+		}
+	},
+
+
+	handleIdentifier: async function (ctx, identifier) {
+		// Identifier
+		try {
+			var translate = new Translate.SearchMultiple();
+			translate.setIdentifier(identifier);
+			let translators = await translate.getTranslators();
+			if (!translators.length) {
+				ctx.throw(501, "No translators available", { expose: true });
+			}
+
+            translate.setTranslator(translators);
+
+			var items = await translate.translate({
+				libraryID: false
+			});
+		}
+		catch (e) {
+			if (e == translate.ERROR_NO_RESULTS) {
+				ctx.throw(501, e, { expose: true });
+			}
+
+			Zotero.debug(e, 1);
+			ctx.throw(
+				500,
+				"An error occurred during translation. "
+					+ "Please check translation with the Zotero client.",
+				{ expose: true }
+			);
+		}
+
+		// Translation can return multiple items (e.g., a parent item and notes pointing to it),
+		// so we have to return an array with keyed items
+		var newItems = [];
+		items.forEach(item => {
```

Comment:
items is declared inside the try-block, so maybe it would be better to move the follow-up processing in there too